### PR TITLE
Add a decorator to HTTP request handlers

### PR DESCRIPTION
**Pull request by mbarnes**
The `Dispatcher`'s `dispatch()` method is doing too much work, and is assuming the entire Commissaire REST API is JSON-RPC based.

The Tang pass-through API that I started working on for https://github.com/projectatomic/commissaire-http/issues/59 would have invalidated that assumption, and I think it's worth addressing even if the Tang API is being shelved for the time being.  My feeling is this will bite us again in the future.

I wrote a `JSONRPC_Handler` class to be used as a decorator for our existing HTTP request handler functions.  The class takes over the pre and post-processing work around the handler call that `dispatch()` was previously doing.

The function renames were motivated by a "don't assume everything is JSON-RPC" mindset.

### CODE REVIEW

**Comment by mbarnes** [comment]
[[multi-host-mgr-2] Refactor to allow for future pass-through APIs](https://trello.com/c/oe5X6qzJ/645-multi-host-mgr-2-refactor-to-allow-for-future-pass-through-apis)

**Comment by ashcrow** [comment]
Looks good so far.

**Comment by mbarnes** [comment]
@ashcrow :arrow_up: Rebased and pushed a couple fixups.  This is ready for review.

**Comment by mbarnes** [comment]
I see two failures; I think one of them preceded the branch.  Looking into it.
```
Failing scenarios:
  features/cluster/hosts.feature:181  Directly setting a cluster host list -- @1.1
  features/cluster/hosts.feature:223  Creating a new host with an invalid cluster name
```

**Comment by mbarnes** [comment]
:arrow_up: Those two fixups take care of the `hosts.feature:181` failure.

**Comment by ashcrow** [comment]
Yeah I think we busted this before this PR.

- ``commissaire-service``: cce65b2ba9b3407a0ccd280a30df138f7949aa72
- ``commissaire-http``:  fe5529e383df909826257a2dccbbd9fcd10ee388 (Current HEAD)
- ``commissaire``: 22b5e889538af9b39ed09128e8b03ffd306be6cd

```
$ ./tools/behave
...
1 feature passed, 11 failed, 0 skipped
10 scenarios passed, 43 failed, 5 skipped
99 steps passed, 43 failed, 204 skipped, 0 undefined
Took 0m12.893s
```

- ``commissaire-service``: cce65b2ba9b3407a0ccd280a30df138f7949aa72
- ``commissaire-http``: c602fd6ebac0c41828a177ce6fa7d482673e7513 (this PR)
- ``commissaire``: 22b5e889538af9b39ed09128e8b03ffd306be6cd

```
$ ./tools/behave
...
1 feature passed, 11 failed, 0 skipped
10 scenarios passed, 43 failed, 5 skipped
99 steps passed, 43 failed, 204 skipped, 0 undefined
Took 0m12.553s
```

**Comment by ashcrow** [comment]
Reviewing and will merge if all looks good. I'll open up a bug for us to figure out why E2E was already failing.

File: src/commissaire_http/dispatcher/__init__.py [hunk 1]
```diff
@@ -210,89 +151,38 @@ def dispatch(self, environ, start_response):
...
-            # but we hijack it for the method that was used over HTTP
-            jsonrpc_msg = {
-                'jsonrpc': '2.0',
-                'id': str(uuid.uuid4()),
-                'method': environ['REQUEST_METHOD'],
-                'params': params,
-            }
 
+        # Add the bus instance to the WSGI environment dictionary.
+        environ['commissaire.bus'] = self._bus
```
Comment (ashcrow) [inline src/commissaire_http/dispatcher/__init__.py]: :+1: I like this pattern of naming in ``environ``.

File: src/commissaire_http/dispatcher/__init__.py [hunk 2]
```diff
@@ -210,89 +151,38 @@ def dispatch(self, environ, start_response):
...
+
+        # Add the routematch results to the WSGI environment dictionary.
+        match_result = self._router.routematch(environ['PATH_INFO'], environ)
+        if match_result is None:
+            start_response(
+                '404 Not Found',
```
Comment (ashcrow) [inline src/commissaire_http/dispatcher/__init__.py]: In the future we should probably start pulling response text from a library or constants.

File: src/commissaire_http/dispatcher/__init__.py [hunk 3]
```diff
@@ -210,89 +151,38 @@ def dispatch(self, environ, start_response):
...
+                [('content-type', 'text/html')])
```
Comment (ashcrow) [inline src/commissaire_http/dispatcher/__init__.py]: Same as above with content-types. Obviously these are not related to your change, just want to record it in a comment for the future.

File: src/commissaire_http/handlers/__init__.py [hunk 4]
```diff
@@ -16,15 +16,188 @@
...
+            else:
+                message = 'Unhandled error code {}'.format(error_code)
+                self.logger.error('{}: {}'.format(message, result))
+                raise Exception(message)
+            start_response(status, [('content-type', 'text/html')])
+            response_body = status[4:]
+
+        elif 'result' in result.keys():
+            status = '200 OK'
+            if environ['REQUEST_METHOD'] == 'PUT':
```
Comment (ashcrow) [inline src/commissaire_http/handlers/__init__.py]: :+1:

**Comment by ashcrow** [comment]
Squashed and merged manually.

File: src/commissaire_http/dispatcher/__init__.py (hunk 2 above)
Comment (mbarnes) [inline src/commissaire_http/dispatcher/__init__.py]: I agree, but I've not yet found a dictionary of such constants.  Neither in `requests` nor Python's Standard Library, which is quite surprising.

File: src/commissaire_http/dispatcher/__init__.py (hunk 1 above)
Comment (mbarnes) [inline src/commissaire_http/dispatcher/__init__.py]: Yeah, that's actually mandated by [PEP 3333](https://www.python.org/dev/peps/pep-3333/#environ-variables):

Finally, the `environ` dictionary may also contain server-defined variables. These variables should be named using only lower-case letters, numbers, dots, and underscores, and should be prefixed with a name that is unique to the defining server or gateway. For example, `mod_python` might define variables with names like `mod_python.some_variable`. 
```

**Comment by mbarnes** [comment]
> Reviewing and will merge if all looks good. I'll open up a bug for us to figure out why E2E was already failing.

Maybe run `vagrant destroy`?  I do that fairly regularly to make sure the test environment is pristine.  The Vagrant environment's been pretty stable for me as of late, though I use the script under `tools` instead of plain `vagrant up`.

**Pull request closed without merging**